feat: Aggregate Transaction View. Changed AggregateBonded signing flow. Inner transaction presentation and Transaction Graphic.

OlegMakarenko opened the pull request:
- Added Aggregate Transaction Details view
- Added support of all transaction types (aggregate inner only)
- Added Transaction Graphic (Inner Transactions tab)
- Added new sign flow for Aggregate Bonded transactions
- Added `(Your address)` label nex to the current address in the TableView
- Improved nested table support in TableView
- Added boolean field to TableView (Checkmark and cross)
- Added compact view to TableView
- Fixed Mosaic field styles

![image](https://user-images.githubusercontent.com/33131259/141784362-95ab9202-8a2a-4a0f-a0fb-2a65ee4dd168.png)

yilmazbahadir reviewed (commented):
Images look great, good work mate.
I put some comments and I'd like to continue after the indentation issues are fixed(ideally via lint:fix script - which should be added as below)
`"lint:fix": "eslint --fix .",`

OlegMakarenko commented on src/components/transaction-graphic/graphics/GraphicComponent.js:
```diff
@@ -0,0 +1,258 @@
+        return `rgb(${color.R},${color.G},${color.B})`;
+    }
+
+    truncString(str, strLen = 4) {
+        if (typeof str === 'string') {
+			if (str.length > strLen * 2)
+				return `${str.substring(0, strLen)}...${str.substring(str.length - strLen, str.length)}`;
+			return str;
+		}
+		console.error('Failed to trunc string. Provided value is not a string');
+		return str;
+    }
```
@yilmazbahadir Sure, I've opened an issue regarding that: https://github.com/symbol/mobile-wallet/issues/265

yilmazbahadir commented:
Probably it'd be also good to get @AnthonyLaw and @bassemmagdy's reviews on this one.

AnthonyLaw commented on src/locales/translations/en.json:
```diff
@@ -432,20 +432,99 @@
+		"endEpoch": "End Epoch",
+		"receivedCosignatures": "Received Cosignatures",
+
+		"v_unlimited": "UNLIMITED",
```
what is the `v_` prefix mean?

OlegMakarenko replied:
It means `value` in the table view

AnthonyLaw commented on src/storage/models/TransactionModel.js:
```diff
@@ -60,3 +67,167 @@ export interface MosaicAliasTransactionModel extends TransactionModel {
+    aliasAction: string;
+}
+
+export interface HashLockModel extends TransactionModel {
```
I believe it missing `lock hash`, you can have a look.

AnthonyLaw commented on src/config/constants.js:
```diff
@@ -0,0 +1,227 @@
+        [ResolutionType.Mosaic]: 'Mosaic',
+    };
+
+    static NetworkType = {
+        [NetworkType.MAIN_NET]: 'MAINNET',
+        [NetworkType.MIJIN]: 'MIJIN',
+        [NetworkType.MIJIN_TEST]: 'MIJIN TESTNET',
+        [NetworkType.TEST_NET]: 'TESTNET',
+        [NetworkType.PRIVATE_TEST]: 'PRIVATE TESTNET',
+        [NetworkType.PRIVATE]: 'PRIVATE',
+    };
```
I think we should only keep `[NetworkType.MAIN_NET]` and `[NetworkType.TEST_NET]` the rest can be remove

AnthonyLaw commented on src/components/transaction-graphic/MosaicAddressRestrictionGraphic.js:
```diff
@@ -0,0 +1,53 @@
+                    height={this.subjectHeight}
+                    address={this.props.signerAddress}
+                />
+                <AccountIcon
+                    x={this.objectPositionX}
+                    y={this.objectPositionY}
+                    width={this.subjectWidth}
+                    height={this.subjectHeight}
+                    address={this.props.targetAddress}
+                />
```
In explorer we are using `MosaicIcon` not sure are you intend to use it.

OlegMakarenko replied:
Yes, I changed that because the target is an address, not a mosaic.

AnthonyLaw replied:
Alright, it should do the same in Explorer as well 👍🏼

OlegMakarenko replied:
Agree, let's revisit it when we have the new design ready

AnthonyLaw commented on src/services/FormatTransaction.js:
```diff
@@ -0,0 +1,745 @@
+                namespace.namespaceId.toHex() ===
+                transaction.namespaceId.toHex()
+        );
+        console.log('namespaceNames', namespaceNames);
```
remove console.log

AnthonyLaw commented on src/components/organisms/transaction/TransactionItem.js:
```diff
@@ -0,0 +1,149 @@
+                </Row>
+                {showDetails && (
+                    <View style={{ paddingVertical: 16 }}>
+                        {this.renderDetails()}
```
does it render anything, it just `return;` ?

AnthonyLaw commented on src/components/organisms/transaction/TransactionItem.js:
```diff
@@ -0,0 +1,149 @@
+                                {this.renderAddress()}
+                            </Text>
+                            <Row style={styles.value} align="center">
+                                {this.renderAction()}
```
does it render anything, it just return; ?

OlegMakarenko replied:
> it just return; 
This is stub method. Cichild classes implement these methods.

Removed this component. Will rework it in the next PR.

AnthonyLaw reviewed (approved):
Great job, all look good 👍🏼 
We need to build a feature branch, for the test team to play around with before merging.

CC: @cryptoBeliever

AnthonyLaw commented:
@OlegMakarenko You may want to disable sonar cloud for the line of code.

https://github.com/symbol/mobile-wallet/blob/5e563b6f80018693273c3f96ef4d9201b6bbb79a/src/store/network.js#L98

yilmazbahadir commented:
@OlegMakarenko Can you please fix the LGTM issues?
https://lgtm.com/projects/g/symbol/mobile-wallet/rev/pr-70333075811b7da820f168663f2d319cb8eb55d9

cryptoBeliever commented:
@yilmazbahadir @OlegMakarenko could you build APK for me? Looks like Jenkins agent is off again.

cryptoBeliever commented:
@OlegMakarenko below my comments

1. For key links (voting, address, VRF, node) and address/mosaic alias there is one icon no matter if this is LINK or UNLINK operation. In explorer for UNLINK we have a strike-through icon, for LINK it's not a strike-through. In a mobile wallet, it's always strike-through not matter if it's LINK or UNLINK.

![image](https://user-images.githubusercontent.com/17684281/142837859-41b59da1-2155-4432-af5a-ded09f639669.png)

2. Transaction Details -> About transaction. For bonded aggregates "Received Cosigntures" are not presented correctly. It's always "n/a". Also in the desktop wallet, this field has: "Co-signatures received" label.

![image](https://user-images.githubusercontent.com/17684281/142837832-57f01e95-52b9-49d1-adaf-3351fabf31cf.png)

3. For mosaic definition transactions there is a missing "divisibility" field.

4. Secret proof/lock and Account metadata. Right side of arrow is not resolved to "You" if the address matches. For example: https://testnet.symbol.fyi/transactions/FEF92C4765139C9E4C89AFE614F9A75A815E59FCE15C4421562C80FEEE718301 

![image](https://user-images.githubusercontent.com/17684281/142838532-799086a2-a75c-412e-8bd2-37c34f0ce683.png)

5. Mosaic global restriction:
- There is missing "Previous restriction value" field
- Restriction types are presented as operators for example =, but in explorer, it's word "Equal"

6. Namespace registration / mosaic creation / has lock / secret locks:
- Would be possible to present also human-readable values instead of only blocks? For example add 30d 0h 0m instead only value 86400?
- it will require additional calls to present the estimated rental fee for mosaic and namespace? If not we can add that.

7. "Mosaic address restriction". In explorer on the right side of arrow, we have "mosaic" icon. In mobile wallet it's target address.

https://testnet.symbol.fyi/transactions/0627321705D4D0424BA7EC796E3A621D85212860A9023C9DD3C78EC6E40884ED

![image](https://user-images.githubusercontent.com/17684281/142839319-44745dff-382d-499a-9f8c-1df2ddb52e55.png)

8. Target address is presented without dashes for mosaic address restriction, account/namespace/mosaic metadata). Same with Account address restriction / multisig modifications and address additions/deletions.

9. Maybe we should change "Waiting for your signature" to Waiting for your review"?

10. I think that we should present transactions unfold by default. The user should review all of them.

OlegMakarenko commented:
@cryptoBeliever 

> For key links (voting, address, VRF, node) and address/mosaic alias there is one icon no matter if this is LINK or UNLINK operation. In explorer for UNLINK we have a strike-through icon, for LINK it's not a strike-through. In a mobile wallet, it's always strike-through not matter if it's LINK or UNLINK.

Fixed.

> Transaction Details -> About transaction. For bonded aggregates "Received Cosigntures" are not presented correctly. It's always "n/a". Also in the desktop wallet, this field has: "Co-signatures received" label.

Fixed. Label text updated.

> For mosaic definition transactions there is a missing "divisibility" field.

Fixed. The `0` values should be diplayed correctly.

> Secret proof/lock and Account metadata. Right side of arrow is not resolved to "You" if the address matches. For example: https://testnet.symbol.fyi/transactions/FEF92C4765139C9E4C89AFE614F9A75A815E59FCE15C4421562C80FEEE718301

Fixed.

> Mosaic global restriction:
> There is missing "Previous restriction value" field

Fixed. The `0` values should be diplayed correctly.

> Restriction types are presented as operators for example =, but in explorer, it's word "Equal"

In the Explorer's Transaction Graphic it also displayed as operators.

> Namespace registration / mosaic creation / has lock / secret locks:
> Would be possible to present also human-readable values instead of only blocks? For example add 30d 0h 0m instead only value 86400?
> it will require additional calls to present the estimated rental fee for mosaic and namespace? If not we can add that.

Sounds good. Let's add it in the next PR (to the Explorer as well).

> "Mosaic address restriction". In explorer on the right side of arrow, we have "mosaic" icon. In mobile wallet it's target address.
>https://testnet.symbol.fyi/transactions/0627321705D4D0424BA7EC796E3A621D85212860A9023C9DD3C78EC6E40884ED

Yes, we should fix it in the Explorer as well.

> Target address is presented without dashes for mosaic address restriction, account/namespace/mosaic metadata). Same with Account address restriction / multisig modifications and address additions/deletions.

Fixed.

> I think that we should present transactions unfold by default. The user should review all of them.

Added.

> Maybe we should allow user to quit here too by adding "Cancel"/"I don't recognize transaction" button? User can close by "x" button but it's not so visible and can force user to just sign because it's easier from UI part.

The extra button will make the purple panel large, and it will be hard to review transaction

cryptoBeliever commented:
@OlegMakarenko 

I.
> Restriction types are presented as operators for example =, but in explorer, it's word "Equal"
In the Explorer's Transaction Graphic it is also displayed as operators.

_I mean here restriction type in explorer it's as text (Equal) but in mobile, it's =_:

https://testnet.symbol.fyi/transactions/B60BBCBF46E8015BF05A344A2D64957AD2C6D6C2F058207119608D3754005E75

![image](https://user-images.githubusercontent.com/17684281/142998554-0f496281-a756-45af-bc63-2b1239637193.png)

II.
> "Mosaic address restriction". In explorer on the right side of arrow, we have "mosaic" icon. In mobile wallet it's target address.
> https://testnet.symbol.fyi/transactions/0627321705D4D0424BA7EC796E3A621D85212860A9023C9DD3C78EC6E40884ED

Yes, we should fix it in the Explorer as well.

 _It's not a mosaic icon in explorer and an address icon in mobile. How it should be presented?_

III.

> I think that we should present transactions unfold by default. The user should review all of them.
Added.

_I meant here to show transaction details by default. Now user has to click to see details of single inner tx. What you think about it?_ 

IV.

> Maybe we should allow user to quit here too by adding "Cancel"/"I don't recognize transaction" button? User can close by "x" button but it's not so visible and can force user to just sign because it's easier from UI part.

The extra button will make the purple panel large, and it will be hard to review transaction

_I'm afraid people will accepting by default_

V.

Maybe we should change "Waiting for your signature" to Waiting for your review"?

_What you think about it?_

cryptoBeliever commented:
Would be good to check on IOS if all looks good. @yilmazbahadir we need to push to TestFlight and ask @NikolaiB to verify.

yilmazbahadir commented:
@cryptoBeliever Jenkins is failing to build the IOS and the android artifacts seem to be corrupted. Will try to do it manually today.
